WiP: Local reply mapper

lukidzi opened the pull request:
As mentioned in #7537 I've created this PR to discuss solution. I haven't added json logs config in this PR because I want to do it as separate task.
Rest of the changes apart from api were created just to check if config is loaded properly(for my usage). 

I would like to start implementation as soon as we create nice and resuable config. 

I need some guidance where should I place feature related files. I thought about either new package `common/local_reply` or `common/http` but I'm open for suggestion.

mattklein123 commented on api/envoy/data/accesslog/v2/BUILD:
```diff
@@ -6,6 +6,7 @@ api_proto_library_internal(
     name = "accesslog",
     srcs = ["accesslog.proto"],
     visibility = [
+        "//envoy/config/filter/network/http_connection_manager/v2:__pkg__",
```
Why is this needed?

lukidzi replied:
I added this because ResponseFlags wasn't visible in proto.

mattklein123 commented on api/envoy/config/filter/network/http_connection_manager/v2/http_connection_manager.proto:
```diff
@@ -450,6 +452,34 @@ message HttpConnectionManager {
+      uint32 status = 1 [(validate.rules).uint32 = {gte: 100, lt: 600}];
+    };
+
+    Match match = 1;
```
I think we should have a repeated set of match and and rewrite rules, and potentially the first match wins? cc @htuch as this relates to our v4 match discussion.

lukidzi replied:
Yes it sound good. We should allow to specify many matchers and rewriters and first match wins.

htuch replied:
Can we see if we can unify with TAP or access log? See my other comment.

mattklein123 commented on api/envoy/config/filter/network/http_connection_manager/v2/http_connection_manager.proto:
```diff
@@ -450,6 +452,34 @@ message HttpConnectionManager {
+    // Configuration which allows to define custom mapping for matched response.
+    message Rewrite {
+      // Definition of status code to which matched one will be mapped.
+      uint32 status = 1 [(validate.rules).uint32 = {gte: 100, lt: 600}];
```
I would model this as an optional WKT, as I think in the future we may very well have other rewrite options such as body replacement, header addition, etc.

mattklein123 commented on api/envoy/config/filter/network/http_connection_manager/v2/http_connection_manager.proto:
```diff
@@ -450,6 +452,34 @@ message HttpConnectionManager {
+    // then response must match both
+    message Match {
+      // Definition of status codes to match.
+      repeated uint32 status_codes = 1;
```
I would model this as effectively a repeated oneof of individual match types, with the final match being a logical AND of all of the conditions.

lukidzi replied:
Changed it

mattklein123 commented on api/envoy/config/filter/network/http_connection_manager/v2/http_connection_manager.proto:
```diff
@@ -450,6 +452,34 @@ message HttpConnectionManager {
+    // Configuration which allows to define custom rules for matching local response.
+    // It works like and operator e.g: if you define status code 404 and body pattern
+    // then response must match both
+    message Match {
```
I really, really, really would like to unify our HTTP match capabilities, but maybe we defer this to v4 as I think any other solution will be too disruptive. @htuch any thoughts here?

htuch replied:
I think that we should tackle this in v4, but meanwhile, for new mathers, can we align them behind either access log or TAP matchers as the basis (even if only subsetted)? I'd prefer not to add yet another matcher format (and surrounding code cruft for parsing and executing match expressions).

lizan replied:
My high level question here is do we really need a match/rewrite semantics here? Since local reply is the output from Envoy (either from core HCM, or extensions), can we change the code of local reply to emit some structured data, and map it to the actual response based on config? A `Rewrite` here seems based on how current local reply works.

lukidzi replied:
I am not sure if we want always to map some response. I thought that it would be nice to just map specifice responses to another one like it has been written in  #7537.

mattklein123 replied:
IMO it's a bit easier to allow the operator to have generic rules for local replies versus having to account for every possible location where a reply is generated, but I could definitely be convinced otherwise.

lizan replied:
I understand the context of #7537, but I'd like this to be generic enough to make it possible to convert all local reply to JSON. Particularly can `Rewrite` be some sort of formatter instead?

`Rewrite` proposed here seems based on how current local reply works (i.e. take what current sendLocalReply send, and rewrite the response header/body), instead can we just make sendLocalReply emit what formatter says for those?

mattklein123 replied:
Yeah agreed this makes sense to rethink rewrite is some combination of both rewriting fields as well as a final format phase?

lukidzi commented:
/wait

lukidzi commented:
/wait

lizan commented:
Tagging @zyfjeff worked on similar thing previously https://github.com/envoyproxy/envoy/pull/6261

mattklein123 commented on api/envoy/config/filter/network/http_connection_manager/v2/http_connection_manager.proto:
```diff
@@ -450,6 +452,42 @@ message HttpConnectionManager {
+    message Match {
+      oneof matcher_specifier{
+        // Definition of status codes to match.
+        google.protobuf.UInt32Value status_codes = 1;
```
Effectively I think this can be a header matcher on `:status`? Per @htuch point elsewhere, is there anything we can do here to share some of this beyond just sharing HeaderMatcher, StringMatcher, etc.?

lukidzi commented:
Ok. Thanks for feedback I'll try to come up with new proposition.

lukidzi commented:
/wait

lukidzi commented:
I've removed changes with matcher/rewrite and added formatter which reuse access_log code to create json format.  Configuration looks same as for  access logs. This is just a concept and I will be glad to hear your opinions.

mattklein123 commented on api/envoy/config/filter/network/http_connection_manager/v2/http_connection_manager.proto:
```diff
@@ -453,6 +453,26 @@ message HttpConnectionManager {
+  // // `HTTP spec <https://tools.ietf.org/html/rfc3986>` and is provided for convenience.
+  // bool merge_slashes = 33;
+
+  // Configuration of local reply send by envoy for http responses.
```
I think I'm missing something. How does this configuration work exactly?

/wait-any

mattklein123 replied:
I see this makes sense and is pretty click. But don't we still then potentially need matching and rewriting as a parallel set of options? cc @ahedberg @lizan for comment

/wait-any

lukidzi replied:
I've added mapper config to allows `response_code` mapping based on `response_flags`. I've reused `AccessLogFilter` to achieve that. Here is sample config which I tested:
```
              local_reply_config:
                mapper:
                  matcher:
                    response_flag_filter:
                      flags:
                        - LH
                        - UH
                  rewriter:
                    status_code: 504
                json_format:
                      foo: "test"
                      timestamp: "%START_TIME(%FT%T.%3fZ)%"
                      level: TRACE
                      logger: envoy
                      request_protocol: "%PROTOCOL%"
                      request_method: "%REQ(:METHOD)%"
                      request_authority: "%REQ(:authority)%"
                      request_path: "%REQ(:PATH)%"
                      response_body: "%RESP_BODY%"
```

lizan replied:
the config looks reasonable, any chance the matcher part can be unified with access logger filter?

for the rewriter part we may need support grpc-status / grpc-message / grpc-status-details-bin, but you may leave that as a TODO for now.

mattklein123 replied:
+1 to merging with the access log filter matching. 

In general though this looks like a great path forward to me.

htuch replied:
+1 as well on merging with existing matchers.

lukidzi replied:
Changed to AccessLogFilter. If it looks ok for you I will try to implement this and if I have problems I will ask for some guidance.

mattklein123 replied:
@lukidzi can you potentially just cleanup this PR and push only the final API and docs? I would like us to agree on that before implementation. Thank you!

ahedberg replied:
Sorry for the delay. I think the current design should let us achieve what I was requesting in #8214. cc @alyssawilk

lukidzi replied:
Added docs and removed rest of changes. If I should be more specific and place them somewhere else I will fix it in next commit. Thanks !

alyssawilk replied:
Yeah, I think this hits 2 things we want - custom error codes and custom error messages.
My one concern is we might have the knobs we need, but if the flags aren't enough maybe we can add in the response details into matcher options.

lukidzi commented:
/wait

htuch commented on api/envoy/config/filter/network/http_connection_manager/v2/http_connection_manager.proto:
```diff
@@ -453,6 +453,61 @@ message HttpConnectionManager {
+  // // setting this option, incoming requests with path `//dir///file` will not match against route
+  // // with `prefix` match set to `/dir`. Defaults to `false`. Note that slash merging is not part of
+  // // `HTTP spec <https://tools.ietf.org/html/rfc3986>` and is provided for convenience.
+  // bool merge_slashes = 33;
```
What's going on here in this commented out block?

lukidzi replied:
My bad. Reverted

lukidzi commented:
/wait

lukidzi commented:
/wait

lukidzi commented:
/wait

lizan commented on api/envoy/config/filter/network/http_connection_manager/v2/http_connection_manager.proto:
```diff
@@ -453,6 +453,40 @@ message HttpConnectionManager {
+
+message ResponseMapper {
+  // Filter is used to determine if the response should be changed.
+  envoy.config.filter.accesslog.v2.AccessLogFilter filter = 1 [(validate.rules).message.required = true];
```
Great, if this works I think adding #next-major-version comment in `AccessLogFilter` to rename it more generically.

htuch replied:
Yep, this sounds great.

htuch commented on api/envoy/config/filter/network/http_connection_manager/v2/http_connection_manager.proto:
```diff
@@ -453,6 +453,40 @@ message HttpConnectionManager {
+  };
+
+  // Configuration of local reply returned by Envoy. Allows to specify mapping and format of response.
+  LocalReplyConfig local_reply_config = 35;
```
Should `LocalReplyConfig` be one-per `ResponseMapper`? What's the tradeoff here?

lukidzi replied:
I think `LocalReplyConfig` should be one by `http_conn_manager`, but we can define many Mappers. The tradeoff is that you can define only one format per manager. So if you want to define custom format for different response it won't be possible.

htuch replied:
If we think that there might be possibly a use case here for multiple custom format, I'd just allow it in the API; the implementation can probably deal with this without too much complexity. I.e. let's be really sure that one format will be sufficient forever if we take the existing design.

lukidzi replied:
I added possibility to define format for each mapper. I think about implementation: if mapper format is defined then use mapper format else use format from config (or default if not defined)

htuch replied:
Given the way you added it, it seems we could punt this until later, if you really want to keep the top-level `reply_format`. So, I'd either rollback that change or remove the top-level `reply_format`; I gather you want it though to reduce verbosity if you have many rewriters?

lukidzi replied:
Yes I thought it might be good idea to define one format at top-level and allow to define custom one for each mapper if someone want different format. I can also remove top level mapper but if someone want to change the format of all responses it will require definition of mapper without `filter/rewriter`.

htuch replied:
OK, given that, maybe just completely ignore the above thread and rollback the per-rewriter bits, since they can be added in a non-breaking way later additively. It'll simplify the implementation work at your end. Sorry about the confusion, the PR clarified how this would look, thanks.

htuch commented on api/envoy/config/filter/network/http_connection_manager/v2/http_connection_manager.proto:
```diff
@@ -453,6 +453,40 @@ message HttpConnectionManager {
+    repeated ResponseMapper mapper = 1;
+
+    // Allows to define custom format of local reply.
+    oneof reply_format {
```
I think something like this belongs in `envoy.type`, as it comes up in other places, e.g. access logs. It's basically a `StringOrJson` type.

lukidzi replied:
I couldn't find this type. Maybe it should be added in v3?

htuch replied:
I'm suggesting we start the trend by adding it in this PR :) Up to you, at least leave a `[#next-major-version:...]` comment if not.

lukidzi replied:
Added new format. You can check if it looks ok. I haven't changed it in v2 accesslog config because it isn't compatible with current acceslog format. Example:
```
format:
  string_format: 
```

lukidzi commented:
/wait

mattklein123 commented on api/envoy/type/format.proto:
```diff
@@ -0,0 +1,20 @@
+message StringOrJson {
+  // Allows to define custom format of returned data.
+  oneof format {
+    // Plain text format.
```
In the final version of the PR can you flesh these docs out to cross-link to the format rules, etc.? It's not super clear how the user is supposed to fill these fields out. Thank you!

/wait

htuch commented on api/envoy/config/filter/network/http_connection_manager/v2/http_connection_manager.proto:
```diff
@@ -456,6 +457,39 @@ message HttpConnectionManager {
+// Configuration of new value for matched local response.
+message ResponseRewriter {
+  // Status code for matched response.
+  google.protobuf.UInt32Value status_code = 1;
```
Can this just be `uint32`?

lukidzi replied:
I think it will be good to keep this field optional if in future someone want to add e.g: headers, body.

htuch commented on api/envoy/config/filter/network/http_connection_manager/v2/http_connection_manager.proto:
```diff
@@ -456,6 +457,39 @@ message HttpConnectionManager {
+
+message LocalReplyConfig {
+  // Configuration of list of mappers which allows to filter and change local response.
+  // Code iterate through mappers until first match of filter.
```
Nit: "The client will iterate through mappers..`

htuch commented on api/envoy/type/format.proto:
```diff
@@ -0,0 +1,44 @@
+    //    protocol: "HTTP/1.1"
+    //    message: "My message"
+    //
+    // The following JSON object would be returned:
```
Can you rephrase the comments in this file to make it generic, i.e. it might be used in request, response, aribtrary config, etc. It's just a way of specifying some data that might be either structured JSON or flat plain text.

lukidzi commented:
/waiting

lukidzi commented:
Do you think it is good time to start the implementation part or configuration requires some changes?

mattklein123 commented:
IMO this API looks very good and I think we can start on the implementation. I think there are more docs that are needed but we can cover that later as part of the main review. @htuch any further fundamental API comments?

/wait

htuch commented:
@lukidzi @mattklein123 API proposal LGTM.
